python/llm/src/bigdl/llm/transformers/qlora.py

```diff
@@ -279,7 +280,9 @@ def prepare_model_for_kbit_training(model, use_gradient_checkpointing=True):
         # cast all non INT8 parameters to fp32
         for param in model.parameters():
             if (param.dtype == torch.float16) or (param.dtype == torch.bfloat16):
-                param.data = param.data.to(torch.float32)
+                # don't cast linear here to avoid lora OOM on arc
+                if param.dim() == 1:
```

Comment:
what does `param.dim() == 1` mean?

Reply:
For `LowbitLinear`, their dtype is torch.uint8 and dim is 1.
For normal `Linear` / `Embedding`, their dtype is torch.bfloat16/torch.float32 and dim is 2.
For `lateynorm`, their dtype is torch.bfloat16/torch.float32 and dim is 1.

Add `param.dim() == 1` to only cast `lateynorm` to fp32 here, otherwise original code will cast `BF16Linear` to fp32, and lora on arc will OOM at this place.

Reply:
> For `LowbitLinear`, their dtype is `torch.uint8` and dim is 1. For normal `Linear` / `Embedding`, their dtype is `torch.bfloat16`/`torch.float32`/`torch.float16` and dim is 2. For `layernorm`, their dtype is `torch.bfloat16`/`torch.float32`/`torch.float16` and dim is 1.
> 
> Add `param.dim() == 1` to only cast `layernorm` to fp32 here, otherwise original code will cast `BF16Linear` to fp32, and lora on arc will OOM at this place.

This is too implementation specific and can be very fragile. Just check if this is a linear layer.

Reply:
> 
> This is too implementation specific and can be very fragile. Just check if this is a linear layer.

Based on `model.parameters()`, we just got every parameter of model, we don't know the parameter from which module so seems can't check if this param is from a linear layer.
Otherwise, we change to like `for module in model.named_modules():`, but it is an iterator over all modules:
```python
model.layers.31.self_attn LlamaAttention(
  (q_proj): LowBitLinear(in_features=4096, out_features=4096, bias=False)
  (k_proj): LowBitLinear(in_features=4096, out_features=4096, bias=False)
  (v_proj): LowBitLinear(in_features=4096, out_features=4096, bias=False)
  (o_proj): LowBitLinear(in_features=4096, out_features=4096, bias=False)
  (rotary_emb): LlamaRotaryEmbedding()
)
model.layers.31.self_attn.q_proj LowBitLinear(in_features=4096, out_features=4096, bias=False)
model.layers.31.self_attn.k_proj LowBitLinear(in_features=4096, out_features=4096, bias=False)
model.layers.31.self_attn.v_proj LowBitLinear(in_features=4096, out_features=4096, bias=False)
model.layers.31.self_attn.o_proj LowBitLinear(in_features=4096, out_features=4096, bias=False)
model.layers.31.self_attn.rotary_emb LlamaRotaryEmbedding()
```
I am still looking an elegant way for this.

Reply:
Do you think change
```python
        for param in model.parameters():
            if (param.dtype == torch.float16) or (param.dtype == torch.bfloat16):
                # don't cast linear here to avoid lora OOM on arc
                if param.dim() == 1:
                    param.data = param.data.to(torch.float32)
```
to below way :
```python
         for module in model.modules():
            if list(module.children()) == []:
                # leaf module
                if not isinstance(module, (torch.nn.Linear, LowBitLinear, torch.nn.Embedding)):
                    for param in module.parameters():
                        if (param.dtype == torch.float16) or (param.dtype == torch.bfloat16):
                            param.data = param.data.to(torch.float32)
```
would be better and acceptable?

Reply:
Yes; BTW, LowBitLinear is still linear

Reply:
Thanks for reminder! Have change this part.